shopfloor_batch_automatic_creation/actions/picking_batch_auto_create.py

```diff
@@ -104,11 +104,17 @@ def _search_pickings(self, picking_types, user=None, commercial_partner=None):
         # We can't use a limit in the SQL search because the 'priority' fields
         # is sometimes empty (it seems the inverse StockPicking.priority field
         # mess up with default on stock.move), we have to sort in Python.
-        return self.env["stock.picking"].search(
+        pickings = self.env["stock.picking"].search(
             self._search_pickings_domain(
                 picking_types, user=user, commercial_partner=commercial_partner
             )
         )
+        # The selected pickings need to match the type of the first picking found.
+        first_picking_type = fields.first(pickings).picking_type_id
+        filtered_pickings = pickings.filtered(
+            lambda p: p.picking_type_id == first_picking_type
+        )
```

Comment:
The dataset can be large, you can't call filtered here. We need a lazy evaluation. Check filter.
In case of group_by_commercial_partner, the second search must respect your new condition also.

Reply:
You can also check if 1 or 2 search needs to be done. Like multiple picking_types or group_by_commercial_partner would require 2 searches. Then you can perform a first one with limit 1 and a second one with the additional domain

Reply:
Can we return a recordset from filter()? From what I see, we can return lists but I'm not sure if there's an effective way to transform it to a recordset.

Regarding the number of checks, if we want to optimize this filter so that it doesn't run more than once in all cases, wouldn't it be better yo simply apply it after the if / else? That way we'd always have batches with one picking type and filter() would only run once.

Reply:
But we'd need a domain for that first one already wouldn't we? What if the first picking doesn't match the user / the allowed picking types?

Reply:
Yes, you need to apply a first domain for the first search.
@lmignon did the same remark in the migration to v16 https://github.com/OCA/wms/pull/534#discussion_r1072404033